sql/core/src/test/scala/org/apache/spark/sql/DataFrameSetOperationsSuite.scala

```diff
@@ -1018,6 +1018,23 @@ class DataFrameSetOperationsSuite extends QueryTest with SharedSparkSession {
     unionDF = df1.unionByName(df2)
     checkAnswer(unionDF, expected)
   }
+
+  test("SPARK-36673: Incorrect Unions of struct") {
+    val df1 = spark.range(2).withColumn("nested", struct(expr("id * 5 AS INNER")))
+    val df2 = spark.range(2).withColumn("nested", struct(expr("id * 5 AS inner")))
```

Comment:
Nice. So, this is also testing case sensitivity?

Reply:
Currently it is always case sensitive for the nested fields in structs during merging (ref. `StructType.merge`). But for `union`, we don't actually want to do a struct merging. For `union`, we don't merge the schema of its children for all cases but resolve columns between them (by position or by name).

Reply:
That's said, for `union`, we don't actually care about case sensitivity as by definition we resolve column by position.